File: experimental/scripts/optimize_mosaic.py
import json
from pathlib import Path
from PIL import Image

BASE_DIR = Path(__file__).resolve().parent.parent
SECTIONS_DIR = BASE_DIR / "assets" / "sections"
OPTIMIZED_DIR = SECTIONS_DIR / "optimized"
MANIFEST_PATH = SECTIONS_DIR / "manifest.json"

MAX_WIDTH = 800
MAX_HEIGHT = 800
QUALITY = 80


def safe_print(message: str) -> None:
    """Print a message, replacing characters the console cannot encode."""
    try:
        print(message)
    except UnicodeEncodeError:
        print(message.encode("ascii", "replace").decode("ascii"))


def optimize_image(input_path: Path, output_path: Path) -> None:
    with Image.open(input_path) as img:
        img = img.convert("RGB")
        img.thumbnail((MAX_WIDTH, MAX_HEIGHT), Image.Resampling.LANCZOS)
        output_path.parent.mkdir(parents=True, exist_ok=True)
        img.save(output_path, "webp", quality=QUALITY, method=6)


def main() -> None:
    with open(MANIFEST_PATH, "r", encoding="utf-8") as f:
        manifest = json.load(f)

    optimized_manifest = []

    for relative_path in manifest:
        input_path = BASE_DIR / relative_path
        # Strip the "assets/sections/" prefix so optimized images mirror the structure
        output_relative = Path(relative_path).relative_to("assets/sections").with_suffix(".webp").as_posix()
        output_path = OPTIMIZED_DIR / output_relative

        try:
            optimize_image(input_path, output_path)
            optimized_entry = f"assets/sections/optimized/{output_relative}"
            optimized_manifest.append(optimized_entry)
            safe_print(f"Optimized: {optimized_entry}")
        except Exception as e:
            safe_print(f"Failed to optimize {relative_path}: {e}")
            optimized_manifest.append(relative_path)

    optimized_manifest_path = SECTIONS_DIR / "manifest.optimized.json"
    with open(optimized_manifest_path, "w", encoding="utf-8") as f:
        json.dump(optimized_manifest, f, indent=2, ensure_ascii=False)

    safe_print(f"\nWrote {optimized_manifest_path} with {len(optimized_manifest)} entries.")


if __name__ == "__main__":
    main()

File: scripts/optimize-mosaic.mjs
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";
import sharp from "sharp";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const SECTIONS_DIR = path.resolve(__dirname, "..", "assets", "sections");
const OPTIMIZED_DIR = path.join(SECTIONS_DIR, "optimized");
const MANIFEST_PATH = path.join(SECTIONS_DIR, "manifest.json");

const MAX_WIDTH = 800;
const MAX_HEIGHT = 800;
const QUALITY = 80;

async function optimize() {
    const manifest = JSON.parse(await fs.promises.readFile(MANIFEST_PATH, "utf8"));
    const optimizedManifest = [];

    for (const relativePath of manifest) {
        const ext = path.extname(relativePath).toLowerCase();
        if (![".jpg", ".jpeg", ".png", ".webp"].includes(ext)) {
            console.log(`Skipping unsupported file: ${relativePath}`);
            optimizedManifest.push(relativePath);
            continue;
        }

        const inputPath = path.resolve(SECTIONS_DIR, relativePath);
        const outputRelativePath = relativePath.replace(/\.[^.]+$/, ".webp");
        const outputPath = path.join(OPTIMIZED_DIR, outputRelativePath);

        try {
            await fs.promises.mkdir(path.dirname(outputPath), { recursive: true });

            await sharp(inputPath)
                .resize({
                    width: MAX_WIDTH,
                    height: MAX_HEIGHT,
                    fit: "inside",
                    withoutEnlargement: true
                })
                .webp({
                    quality: QUALITY,
                    effort: 4,
                    nearLossless: false
                })
                .toFile(outputPath);

            const rel = "assets/sections/optimized/" + outputRelativePath.replace(/\\/g, "/");
            optimizedManifest.push(rel);
            console.log(`Optimized: ${rel}`);
        } catch (error) {
            console.error(`Failed to optimize ${relativePath}:`, error.message);
            optimizedManifest.push(relativePath);
        }
    }

    await fs.promises.writeFile(
        path.join(SECTIONS_DIR, "manifest.optimized.json"),
        JSON.stringify(optimizedManifest, null, 2)
    );

    console.log(`\nWrote manifest.optimized.json with ${optimizedManifest.length} entries.`);
}

optimize().catch((error) => {
    console.error(error);
    process.exit(1);
});
